### PR TITLE
Added serialization of irept and dstringt

### PR DESCRIPTION
**Pull request by NathanJPhillips**
This adds a generic serialization framework that is used by summary virtualization in the security scanner.
It allows for targets including storage as JSON or using a database server.
It uses a non-virtual serialize method (through the use of templates) that avoids adding size to the objects to be serialized.
A trade-off was the choice to use single function serialization (rather than separate load/save functions). This minimises the impact on classes to be serialized but does sometimes cause issues as const objects can't naively be serialized out.
make_valid_filename has been added. This is for use in JSON serialization. While this serialization implementation is not included in the CBMC repository, this repository seems the right place for that function.
The serialization of irept in particular is based on a previous save to JSON function used by the security scanner and does not necessarily do a full serialization. If this is to be used in other contexts then summary virtualization, it could easily be extended (preferably by someone who understands the internals of irept better).

### CODE REVIEW

File: src/util/serializer.h [hunk 1]
```diff
@@ -0,0 +1,818 @@
+#pragma once
```
Comment (peterschrammel) [inline src/util/serializer.h]: Remove. We use header guards.

File: src/util/serializer.h [hunk 2]
```diff
@@ -0,0 +1,818 @@
...
+
```
Comment (peterschrammel) [inline src/util/serializer.h]: Comment block missing, see http://cbmc-linter.diffblue.com.s3-website-us-east-1.amazonaws.com/linter-07d33070134a7821cf8dfc6608e2791737df54db.txt

File: src/util/serializer.h [hunk 3]
```diff
@@ -0,0 +1,818 @@
...
+#include <string>
+#include <functional>
+#include <vector>
+#include <set>
+#include <map>
+#ifdef USE_BOOST
+#include <boost/bimap.hpp>
+#endif
+
+
```
Comment (peterschrammel) [inline src/util/serializer.h]: remove two blank lines

File: src/util/serializer.h [hunk 4]
```diff
@@ -0,0 +1,818 @@
...
+  Outputs:
+
+  Purpose:
+    Creates a serializert for a sub-object of a larger serialization.
+    This is only called internally by sub-class implementations.
+
+  \*******************************************************************/
+  serializert(serializert *parent, bool is_read)
+    : parent(parent), is_read(is_read)
```
Comment (peterschrammel) [inline src/util/serializer.h]: see http://cbmc-linter.diffblue.com.s3-website-us-east-1.amazonaws.com/linter-07d33070134a7821cf8dfc6608e2791737df54db.txt

(several occurrences)

File: src/util/serializer.h [hunk 5]
```diff
@@ -0,0 +1,818 @@
...
+      one of its parents
+
+  Purpose:
+    Gets traits attached to this serializer of a specific type.
+
+  \*******************************************************************/
+  template<typename traitst>
+  traitst &get_traits() const
+  {
+    traitst * result = dynamic_cast<traitst *>(traits);
```
Comment (peterschrammel) [inline src/util/serializer.h]: *result=dyn...

File: src/util/serializer.h [hunk 6]
```diff
@@ -0,0 +1,818 @@
...
+    if(result!=nullptr)
+      return *result;
+    if(parent==nullptr)
+      throw std::logic_error(
```
Comment (peterschrammel) [inline src/util/serializer.h]: for uniformity, use assert()

File: src/util/serializer.h [hunk 7]
```diff
@@ -0,0 +1,818 @@
...
+      If there were already traits attached to this serializer
+
+  Purpose:
+    Sets traits attached to this serializer.
+
+  \*******************************************************************/
+  void set_traits(serializer_traitst &traits)
+  {
+    if(this->traits!=nullptr)
+      throw std::logic_error(
```
Comment (peterschrammel) [inline src/util/serializer.h]: for uniformity, use assert()

File: src/util/serializer.h [hunk 8]
```diff
@@ -0,0 +1,818 @@
...
+      std::vector<wrappert> wrapper_vector;
+      for(typename mapt::reference elt : field)
+        wrapper_vector.emplace_back(elt);
+      serializable_obj_collection_writert<std::vector<wrappert>>
+        writer(wrapper_vector);
+      write_array(name, writer);
+    }
+  }
+
+public:
```
Comment (peterschrammel) [inline src/util/serializer.h]: put public section first so that I don't have to scroll down to see the interface of the class

File: src/util/serializer.h [hunk 9]
```diff
@@ -0,0 +1,818 @@
...
+  \*******************************************************************/
+  explicit serializert(bool is_read)
+    : parent(nullptr), is_read(is_read)
+  {
+  }
+
+  // The virtual destructor ensures sub-classes are disposed correctly.
+  virtual ~serializert() = default;
+
+public:
```
Comment (peterschrammel) [inline src/util/serializer.h]: merge public sections (resp. protected, private) so that it is easier to see which methods are in the interface of the class

File: src/util/serializer.h [hunk 10]
```diff
@@ -0,0 +1,818 @@
...
+    name: The name of the field to serialize.
+    field: The field to serialize.
+
+  Outputs:
+
+  Purpose:
+    Serializes a field.
+
+  \*******************************************************************/
+  virtual void serialize(const char *name, bool &field) = 0;
```
Comment (peterschrammel) [inline src/util/serializer.h]: put a blank line before and after declaration

File: src/util/serializer.h [hunk 11]
```diff
@@ -0,0 +1,818 @@
...
+  Purpose:
+    Sets traits attached to this serializer.
+
+  \*******************************************************************/
+  void set_traits(serializer_traitst &traits)
+  {
+    if(this->traits!=nullptr)
+      throw std::logic_error(
+        "Tried to set traits twice on the same serializert");
+    this->traits = &traits;
```
Comment (peterschrammel) [inline src/util/serializer.h]: =&

File: src/util/serializer.h [hunk 12]
```diff
@@ -0,0 +1,818 @@
...
+    std::logic_error:
+      If there were already traits attached to this serializer
+
+  Purpose:
+    Sets traits attached to this serializer.
+
+  \*******************************************************************/
+  void set_traits(serializer_traitst &traits)
+  {
+    if(this->traits!=nullptr)
```
Comment (peterschrammel) [inline src/util/serializer.h]: It's more common in the codebase to use a parameter _traits and refer to the member without this.

File: src/util/dstring.cpp [hunk 13]
```diff
@@ -7,3 +7,33 @@ Author: Daniel Kroening, kroening@kroening.com
...
+    Serializes this instance to/from the given serializer.
+
+\*******************************************************************/
+void dstringt::serialize(serializert &serializer)
+{
+  if(serializer.is_for_reading())
+  {
+    std::string str;
+    serializer.serialize("dstring", str);
+    no = string_container[str];
```
Comment (peterschrammel) [inline src/util/dstring.cpp]: whole PR: remove whitespace around =

File: src/util/serializer.h [hunk 14]
```diff
@@ -0,0 +1,818 @@
...
+  //  method
+  template<typename collectiont>
+  class serializable_obj_collection_writert : public collection_writert
+  {
+  public:
+    typedef typename collectiont::iterator iteratort;
+    typedef typename collectiont::reference referencet;
+
+  private:
+    iteratort it,
```
Comment (peterschrammel) [inline src/util/serializer.h]: use two declarations

File: src/util/serializer.h [hunk 15]
```diff
@@ -0,0 +1,818 @@
...
+  //  value field)
+  template<typename collectiont>
+  class serializable_collection_writert : public collection_writert
+  {
+  public:
+    typedef typename collectiont::iterator iteratort;
+    typedef typename collectiont::reference referencet;
+
+  private:
+    iteratort it,
```
Comment (peterschrammel) [inline src/util/serializer.h]: use two declarations

File: src/util/dstring.h [hunk 16]
```diff
@@ -13,6 +13,9 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include "string_container.h"
 
+class serializert;
+
```
Comment (peterschrammel) [inline src/util/dstring.h]: remove blank lines

File: src/util/serializer.h [hunk 17]
```diff
@@ -0,0 +1,818 @@
...
+    }
+
+    // This is an implicit conversion back to pairt
+    operator pairt &() const { return value; }
+
+    void serialize(serializert &serializer)
+    {
+      serializer.serialize(
+        "first",
+        const_cast<typename std::remove_const<
```
Comment (peterschrammel) [inline src/util/serializer.h]: Why is this needed?

File: src/util/serializer.h (hunk 6 above)
Comment (NathanJPhillips) [inline src/util/serializer.h]: In library functions it is better to throw even in release mode when the library is used wrongly. In code where you have control of the caller and can be sure that it should never be used wrongly then an assert (that is compiled away when NDEBUG is set) is sufficient.

File: src/util/serializer.h (hunk 7 above)
Comment (NathanJPhillips) [inline src/util/serializer.h]: As above. assert does something different.

File: src/util/serializer.h (hunk 11 above)
Comment (NathanJPhillips) [inline src/util/serializer.h]: Won't that be likely to be confused with a single operator (a la &=)?

File: src/util/dstring.h (hunk 16 above)
Comment (NathanJPhillips) [inline src/util/dstring.h]: Are you making up new coding standards just to make things less structured/readable?

File: src/util/serializer.h (hunk 17 above)
Comment (NathanJPhillips) [inline src/util/serializer.h]: See the purpose section of the comment block for this class.

File: src/util/serializer.h (hunk 9 above)
Comment (NathanJPhillips) [inline src/util/serializer.h]: On a class this complicated, moving the helper functions all together away from the code they help will make things more confusing, not less. The class is grouped into sections, if anything I would suggest putting comments at the head of each section might be a better way of making things clearer.

File: src/util/serializer.h (hunk 8 above)
Comment (NathanJPhillips) [inline src/util/serializer.h]: See comment above.

File: src/util/dstring.h (hunk 16 above)
Comment (peterschrammel) [inline src/util/dstring.h]: One blank line should be enough to visually separate blocks. Two, three or more blank lines just stretch the code unnecessarily.

File: src/util/serializer.h [hunk 18]
```diff
@@ -0,0 +1,845 @@
...
+  public:
+    typedef typename collectiont::iterator iteratort;
+    typedef typename collectiont::reference referencet;
+
+  private:
+    iteratort it;
+    iteratort end_it;
+
+  public:
+    explicit serializable_obj_collection_writert(collectiont & collection)
```
Comment (peterschrammel) [inline src/util/serializer.h]: &collection

File: src/util/serializer.h [hunk 19]
```diff
@@ -0,0 +1,845 @@
...
+  public:
+    typedef typename collectiont::iterator iteratort;
+    typedef typename collectiont::reference referencet;
+
+  private:
+    iteratort it;
+    iteratort end_it;
+
+  public:
+    explicit serializable_collection_writert(collectiont & collection)
```
Comment (peterschrammel) [inline src/util/serializer.h]: &collection

File: src/util/serializer.h [hunk 20]
```diff
@@ -0,0 +1,845 @@
...
+    const char *name,
+    std::function<void(serializert &serializer)> read_elt) = 0;
+  virtual void write_array(
+    const char *name, collection_writert &collection_writer) = 0;
+
+public:
+  // A helper class for writing vectors of objects that implement a serialize
+  //  method
+  template<typename collectiont>
+  class serializable_obj_collection_writert : public collection_writert
```
Comment (peterschrammel) [inline src/util/serializer.h]: https://github.com/diffblue/cbmc/blob/master/CODING_STANDARD#L29

File: src/util/serializer.h [hunk 21]
```diff
@@ -0,0 +1,845 @@
...
+    typedef typename collectiont::iterator iteratort;
+    typedef typename collectiont::reference referencet;
+
+  private:
+    iteratort it;
+    iteratort end_it;
+
+  public:
+    explicit serializable_obj_collection_writert(collectiont & collection)
+      : it(collection.begin()), end_it(collection.end())
```
Comment (peterschrammel) [inline src/util/serializer.h]: https://github.com/diffblue/cbmc/blob/master/CODING_STANDARD#L31

File: src/util/serializer.h (hunk 9 above)
Comment (peterschrammel) [inline src/util/serializer.h]: ok

File: src/util/serializer.h (hunk 17 above)
Comment (peterschrammel) [inline src/util/serializer.h]: ok

File: src/util/serializer.h [hunk 22]
```diff
@@ -0,0 +1,845 @@
...
+  Purpose:
+    Sets traits attached to this serializer.
+
+  \*******************************************************************/
+  void set_traits(serializer_traitst &serializer_traits)
+  {
+    if(traits!=nullptr)
+      throw std::logic_error(
+        "Tried to set traits twice on the same serializert");
+    traits = &serializer_traits;
```
Comment (peterschrammel) [inline src/util/serializer.h]: traits=&serializer_traits

File: src/util/serializer.h [hunk 23]
```diff
@@ -0,0 +1,845 @@
...
+      If there were already traits attached to this serializer
+
+  Purpose:
+    Sets traits attached to this serializer.
+
+  \*******************************************************************/
+  void set_traits(serializer_traitst &serializer_traits)
+  {
+    if(traits!=nullptr)
+      throw std::logic_error(
```
Comment (peterschrammel) [inline src/util/serializer.h]: Exceptions should be used to handle errors that should be caught and reported to the user. Typically, this concerns errors that are due to erroneous user input. There is no point in catching a programming error. This might be abused to hide programming errors by catch(...) {}. The program should abort on a programming error. Therefore we prefer assert() in such situations.

File: src/util/serializer.h (hunk 8 above)
Comment (peterschrammel) [inline src/util/serializer.h]: ok

File: src/util/serializer.h [hunk 24]
```diff
@@ -0,0 +1,845 @@
...
+    name: The name of the field to serialize.
+    field: The field to serialize.
+
+  Outputs:
+
+  Purpose:
+    Serializes a field.
+
+  \*******************************************************************/
+  virtual void serialize(const char *name, char &field) = 0;
```
Comment (peterschrammel) [inline src/util/serializer.h]: )=0; (many occurrences)

File: src/util/serializer.h [hunk 25]
```diff
@@ -0,0 +1,845 @@
...
+  This is used to augment the behaviour of a family of serializers, typically
+  by passing data that is required for serializing a particular type of object
+  through to a sub-object serialization routine.
+
+\*******************************************************************/
+class serializer_traitst
+{
+public:
+  // The virtual destructor ensures sub-classes are disposed correctly.
+  virtual ~serializer_traitst() = default;
```
Comment (peterschrammel) [inline src/util/serializer.h]: `)=default;` or `) {}`

File: src/util/dstring.h (hunk 16 above)
Comment (NathanJPhillips) [inline src/util/dstring.h]: Your argument sounds like it could be used to say we only need one type of header in documents. Different things need different levels of separation. I'll implement this because I'm giving up trying to make nice code on this project but I wish I could have made things better here.

File: src/util/serializer.h (hunk 21 above)
Comment (NathanJPhillips) [inline src/util/serializer.h]: But then it won't fit in the 80 character limit.

File: src/util/serializer.h (hunk 23 above)
Comment (NathanJPhillips) [inline src/util/serializer.h]: Giving up arguing and implementing this.

**Comment by NathanJPhillips** [comment]
@smowton unfortunately if we do that it will mess up the generated documentation (when we finally get around to producing it). Each function needs documenting so that HTML documentation won't have missing pages and so that the IDE will be able to give information on parameters for each overload.
